File: customer-java-webflux/src/main/java/com/prez/config/CustomerWebServiceConfig.java
package com.prez.config;

import static org.springframework.http.HttpHeaders.ACCEPT;
import static org.springframework.http.HttpHeaders.CONTENT_TYPE;
import static org.springframework.http.MediaType.APPLICATION_JSON_VALUE;

import brave.SpanCustomizer;
import com.prez.lib.tracing.ResponseMarkerFilter;
import com.prez.lib.tracing.SpanCustomizationWebClientFilter;
import com.prez.ws.CustomerWSClient;
import com.prez.ws.CustomerWSProperties;
import io.micrometer.core.instrument.DistributionSummary;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.boot.context.properties.EnableConfigurationProperties;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.web.reactive.function.client.WebClient;

@Configuration
@EnableConfigurationProperties(CustomerWSProperties.class)
public class CustomerWebServiceConfig {


  @Bean
  public CustomerWSClient customerWsClient(CustomerWSProperties properties,
                                           @Qualifier("customerWSSummary") DistributionSummary customerWSSummary,
                                           SpanCustomizer spanCustomizer) {
    WebClient webClient = WebClient.builder()
        .defaultHeader(ACCEPT, APPLICATION_JSON_VALUE)
        .defaultHeader(CONTENT_TYPE, APPLICATION_JSON_VALUE)
        .filter(new SpanCustomizationWebClientFilter("CustomerWS", spanCustomizer))
        .filter(new ResponseMarkerFilter(customerWSSummary))
        .build();

    return new CustomerWSClient(properties, webClient);
  }
}

File: customer-kotlin-coroutines-controllers/src/main/kotlin/com/prez/config/CustomerWebServiceConfig.kt
package com.prez.config

import brave.SpanCustomizer
import com.prez.lib.tracing.ResponseMarkerFilter
import com.prez.lib.tracing.SpanCustomizationWebClientFilter
import com.prez.ws.CustomerWSClient
import com.prez.ws.CustomerWSProperties
import io.micrometer.core.instrument.DistributionSummary
import org.springframework.boot.context.properties.ConfigurationProperties
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.http.HttpHeaders.ACCEPT
import org.springframework.http.HttpHeaders.CONTENT_TYPE
import org.springframework.http.MediaType.APPLICATION_JSON_VALUE
import org.springframework.web.reactive.function.client.WebClient
import javax.validation.constraints.NotBlank

@Configuration
@ConfigurationProperties(prefix = "customer.ws")
class CustomerWebServiceConfig {

  @NotBlank
  lateinit var url: String

  @Bean
  internal fun CustomerWebServiceConfig(): CustomerWSProperties {
    return CustomerWSProperties(
      url = url
    )
  }

  @Bean
  internal fun customerWsClient(
    properties: CustomerWSProperties,
    customerWSSummary: DistributionSummary,
    spanCustomizer: SpanCustomizer
  ): CustomerWSClient {
    val webClient = WebClient.builder()
      .defaultHeader(ACCEPT, APPLICATION_JSON_VALUE)
      .defaultHeader(CONTENT_TYPE, APPLICATION_JSON_VALUE)
      .filter(SpanCustomizationWebClientFilter("CustomerWS", spanCustomizer))
      .filter(ResponseMarkerFilter("CustomerWS", customerWSSummary))
      .build()

    return CustomerWSClient(properties, webClient)
  }
}
